### ISSUE TITLE
Support GitLab url with custom path

### ISSUE THREAD

**Issue by JIoJIaJIu** [iterative/cml#463]
When gitlab is exposed on the path `https://mydomain.com/gitlab` cml cannot work with such types of urls - the generated api endpoint is `https://mydomain.com/api/v4` but should be `https://mydomain.com/gitlab/api/v4` 

Also it can be appropriate to support custom ports in the urls

**Comment by 0x2b3bfa0** [comment]
> _Also it can be appropriate to support custom ports in the urls_

As per [the `url` documentation](https://nodejs.org/api/url.html#url_class_url), we already support ports out of the box through the use of the `host` property instead of `hostname` for connections. Please let us know if it doesn't work correctly.

**Comment by 0x2b3bfa0** [comment]
@JIoJIaJIu, we can't reliably determine whether a path component is part of the base address or just belongs to a specific repository path, because the latter could be structured inside [subgroups](https://about.gitlab.com/solutions/subgroups) with a variable number of path components:

    https://gitlab.local:443/base-path-component/user-or-group/first-subgroup/second-subgroup/project

**Comment by 0x2b3bfa0** [comment]
The most elegant solution I can think of would involve allowing users to specify the base and the repository path separately:

```console
    cml-runner \
        --base "https://gitlab.local:443/base-path-component" \
        --repo "user-or-group/first-subgroup/second-subgroup/project"
```

We could keep backwards compatibility by requiring either both `--base` and `--repo` or just the latter if it matches `/^https?:\/\//` or similar.

**Comment by DavidGOrtega** [comment]
@0x2b3bfa0 subgroups should be supported as far as I remember 🤔

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, they are, but only when the GitLab root is the server root. When using `https://mydomain.com/gitlab` or similar, we don't have any way to determine if the `/gitlab` path component is part of the path or is the server root.

**Comment by DavidGOrtega** [comment]
I see, well this is a very specific set... I can imagine a solution... we could traverse the path asking for the api endpoint until we have it

**Comment by 0x2b3bfa0** [comment]
I thought the same, but it seemed a bit hacky. Anyhow, it might actually be a good solution for this specific condition. We can skip the last two path components and start from the root in order to optimize algorithm for the best case.

```javascript
const path = '/gitlab/group/subgroup/user/project';
const components = path.split('/').filter(Boolean);

for (let index = 0; index <= components.length - 2; index++)
    console.log(components.slice(0, index).join('/'));
```

**Comment by DavidGOrtega** [comment]
> We can skip the last two path components and start from the root in order to optimize algorithm for the best case

No really needed if the url is a proper gl setup it will break before the proposed condition. It could be all promisified to test all possible endpoints at once.

**Comment by DavidGOrtega** [comment]
What is the remote.origin.url that get @JIoJIaJIu and @0x2b3bfa0 anyone knows?

**Comment by 0x2b3bfa0** [comment]
> _No really needed if the url is a proper gl setup it will break before the proposed condition._
Agreed: the suggested optimization would only matter for broken links, where we could return earlier.

> _It could be all promisified to test all possible endpoints at once._
That would be infinitely better, but would require the `request` function to accept a custom `api_v4` variable on each invocation.

**Comment by 0x2b3bfa0** [comment]
> _What is the remote.origin.url that get @JIoJIaJIu and @0x2b3bfa0 anyone knows?_

❓

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, after a head-scratching session, I have been unable of using a promise to run every operation in parallel, because:

1. The class [`constructor`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/constructor) function must be synchonous.
2. [`Promise.all`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) would require a blocking [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await) that is not possible.

Anyhow, ECMAScript is not my mother tongue and I might be missing something important. Do you know any elegant way of running asynchronous operations on a constructor without requiring users to call an additional initialization function or moving the awaiting code to methods?

**Comment by 0x2b3bfa0** [comment]
Nevermind: solved through #464 with a bit of [memoization](https://en.wikipedia.org/wiki/Memoization) and calls on each function (argh!).

**Comment by 0x2b3bfa0** [comment]
🔔  @JIoJIaJIu, it should be fixed as soon as the fix gets released; it should only take a few minutes.
